### ISSUE TITLE
Label predictions very slow, especially with large (>600) label sets 

### ISSUE THREAD

**Issue by shashi-netra** [neuml/txtai#227]
```
def get_text_classifications(text, ref_labels):
    text_dict={}
    text_dict  = labels(text, ref_labels) 
    return text_dict
```
```
text = "Game of Thrones Season 7: Long Walk Official Promo REACTION ''DISCUSSION!""
ref_labels = [...600+ items long...]
preds = get_text_classifications(text,ref_labels)
pp.pprint(preds)

```
This code takes almost a minute to run, is that expected behavior? How can I improve the performance for zero-shot classification with large label sets? I am running this on a CPU.

```
~ # cat /proc/cpuinfo | grep "model name" 
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
model name	: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
```

Thanks so much for a wonderful project. I could give this project multiple stars if I could!

**Comment by davidmezzetti** [comment]
Thank you for the kind words! I'm glad txtai is helping you with your work.

Regarding performance, while I expect it to run slower on the CPU, a minute for a single text record does seem long. But for 600 labels, it effectively has to do 600 operations. 

Under the covers, a zero-shot MNLI model tokenizes text and labels like this when running model queries.

```
text|label1
text|label2
text|label3
...
```

Couple different options I'd look at to improve performance:

1. Use a GPU - easy idea but that might not be feasible especially if running on cloud compute
2. Pick a different [MNLI model](https://huggingface.co/models?search=mnli). There are smaller models than the default bart-large-mnli which is a large model. There will be a tradeoff on performance though.
3. Use Torch quantization. This is enabled by setting quantize=True in the Labels() constructor.
4. Export the model to ONNX as there are some optimizations to be gained there. [Example notebook on that](https://colab.research.google.com/github/neuml/txtai/blob/master/examples/18_Export_and_run_models_with_ONNX.ipynb). 
4. The Labels object also works with standard text classification models. One thing that has worked well for me is using a zero-shot model to build a training set and then training a standard classification model for that. [An example notebook doing that is here](https://colab.research.google.com/github/neuml/txtai/blob/master/examples/17_Train_without_labels.ipynb).

I think option 5 is the best bet. Another advantage of this is the training can be done locally on a GPU system. Probably wouldn't take building a ton of training data. I've seen decent results with just a couple hundred or thousand rows. If the labels are independent, it's even easier to build abundant data as a single input could produce multiple training rows.
